### PR TITLE
wip: N-D `LinearOperator` and `cg`

### PR DESCRIPTION
**Pull request by lucascolley**
Building on https://github.com/scipy/scipy/pull/23836

https://github.com/kokkos/kokkos-kernels/blob/develop/batched/sparse/impl/KokkosBatched_CG_Team_Impl.hpp

https://ieeexplore.ieee.org/document/10054414 section VI

### CODE REVIEW

File: scipy/sparse/linalg/_isolve/iterative.py [hunk 1]
```diff
@@ -399,23 +401,40 @@ def cg(A, b, x0=None, *, rtol=1e-5, atol=0., maxiter=None, M=None, callback=None
...
-            beta = rho_cur / rho_prev
-            p *= beta
+            beta = xpx.apply_where(
+                not converged,
+                (rho_cur, rho_prev),
+                lambda cur, prev: cur / prev,
+                fill_value=0.0,
+                xp=np
+            )
+            p = (beta * p.T).T
```
Comment (lucascolley) [inline scipy/sparse/linalg/_isolve/iterative.py]: @izaid this is the proposed implementation of 'masking' to match what the paper is doing for `alpha` and `beta`:

> While waiting for the slowest solution to converge, all the previously converged solutions are not updated anymore; their corresponding steps α and β are set to zero. This stopped update allows to enforce that no overflow, underflow, or NaN is generated when dealing with non-zero vectors which have zero norm due to rounding errors.

Image

**Comment by lucascolley** [comment]
Prototype working with `array_api_strict` and `jax.numpy`! (no JIT yet)

```python
In [1]: import array_api_strict as xp; from scipy.sparse.linalg import cg, LinearOperator; import numpy as np

In [2]: def solve(N, batch, report_index=0, batched=False):
   ...:     rng = np.random.default_rng(0)
   ...:     M = rng.standard_normal((N, N))
   ...:     M = xp.asarray(M)
   ...:     reg = 1e-3
   ...:
   ...:     if batched:
   ...:         M = xp.broadcast_to(M[xp.newaxis, ...], (batch, *M.shape))
   ...:
   ...:     def matvec(x):
   ...:         return xp.squeeze(M.mT @ (M @ x[..., xp.newaxis]), axis=-1) + reg * x
   ...:
   ...:     shape = (batch, N, N) if batched else (N, N)
   ...:     A = LinearOperator(shape, matvec=matvec, dtype=xp.float64, xp=xp)
   ...:
   ...:     b = rng.standard_normal(N)
   ...:     b = xp.asarray(b)
   ...:
   ...:     if batched:
   ...:         b = xp.reshape(xp.arange(batch, dtype=xp.float64), (batch, 1)) * b
   ...:         x, info = cg(A, b, atol=1e-8, maxiter=5000)
   ...:         assert info == 0
   ...:         print(f"{x[report_index, ...]}")
   ...:     else:
   ...:         for i in xp.arange(batch, dtype=xp.float64):
   ...:             x, info = cg(A, i*b, atol=1e-8, maxiter=5000)
   ...:             assert info == 0
   ...:             if i == report_index:
   ...:                 print(x)
   ...:

In [3]: solve(5, 10, report_index=7)
Array([ 10.91985197,  -5.53737923,
        -6.96397906, -35.6473016 ,
        13.48931722], dtype=array_api_strict.float64)

In [4]: solve(5, 10, report_index=7, batched=True)
Array([ 10.91985197,  -5.53737923,
        -6.96397906, -35.6473016 ,
        13.48931722], dtype=array_api_strict.float64)

In [5]: import jax
   ...: jax.config.update("jax_enable_x64", True)

In [6]: import jax.numpy as xp

In [7]: solve(5, 10, report_index=7, batched=True)
[ 10.91985197  -5.53737923  -6.96397906 -35.6473016   13.48931722]
```

For JIT I think we will need to do some fancy registration of the linear operator classes as PyTrees, along the lines of https://docs.jax.dev/en/latest/_autosummary/jax.tree_util.register_pytree_node.html#jax.tree_util.register_pytree_node.

File: scipy/sparse/linalg/_interface.py [hunk 2]
```diff
@@ -214,19 +216,24 @@ def _matmat(self, X):
         define matrix multiplication (though in a very suboptimal way).
         """
 
-        return np.hstack([self.matvec(col.reshape(-1,1)) for col in X.T])
+        # X.mT here?
+        return np.hstack([self.matvec(col.reshape(-1, 1)) for col in X.T])
```
[the remark this answers is not in the document]
Comment (lucascolley) [reply scipy/sparse/linalg/_interface.py]: @izaid to maintain compatibility with user-defined `matvec` functions which do not support batching, I think we do need to keep this hstack.

I think it could be done all in one go for batched matvec functions. But I think that would have to be opt-in in the API.
